checks.d/vsphere.py

```diff
@@ -523,112 +523,181 @@ def get_external_host_tags(self):
 
         return external_host_tags
 
-    @atomic_method
-    def _cache_morlist_raw_atomic(self, i_key, obj_type, obj, tags, regexes=None, include_only_marked=False):
-        """ Compute tags for a single node in the vCenter rootFolder
-        and queue other such jobs for children nodes.
-        Usual hierarchy:
-        rootFolder
-            - datacenter1
-                - compute_resource1 == cluster
-                    - host1
-                    - host2
-                    - host3
-                - compute_resource2
-                    - host5
-                        - vm1
-                        - vm2
-        If it's a node we want to query metric for, queue it in self.morlist_raw
-        that will be processed by another job.
+    def _discover_mor(self, instance_key, obj, tags, regexes=None, include_only_marked=False):
         """
-        ### <TEST-INSTRUMENTATION>
-        t = Timer()
-        self.log.debug("job_atomic: Exploring MOR {0} (type={1})".format(obj, obj_type))
-        ### </TEST-INSTRUMENTATION>
-        tags_copy = deepcopy(tags)
+        Explore vCenter infrastructure to discover hosts, virtual machines
+        and compute their associated tags.
+
+
+        Start with the vCenter `rootFolder` and proceed recursively,
+        queueing other such jobs for children nodes.
+
+        Example topology:
+            ```
+            rootFolder
+                - datacenter1
+                    - compute_resource1 == cluster
+                        - host1
+                        - host2
+                        - host3
+                    - compute_resource2
+                        - host5
+                            - vm1
+                            - vm2
+            ```
+
+        If it's a node we want to query metric for, queue it in `self.morlist_raw` that
+        will be processed by another job.
+        """
+        @atomic_method
+        def browse_mor(obj, prev_tags, depth):
```

Comment:
way more flexible, the less assumptions about the resource hierarchy the better! 💯